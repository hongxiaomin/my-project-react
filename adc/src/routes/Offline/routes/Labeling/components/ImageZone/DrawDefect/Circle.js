import React from 'react'
import PropTypes from 'prop-types'

export class Circle extends React.Component {
  constructor() {
    super()

    this.circle = null
    this.down = false
    this.startInfo = {
      x: 0,
      y: 0,
    }
    this.state = {

    }
  }

  componentDidMount() {
    this.circle.addEventListener('mousedown', (e) => {
      this.down = true

      this.startInfo = {
        x: e.offsetX,
        y: e.offsetY,
      }

      e.stopPropagation()
      e.preventDefault()
    })
    window.addEventListener('mouseup', () => {
      if (!this.down) return
      this.down = false
      this.props.end()
    })
    window.addEventListener('mousemove', (e) => {
      const nodeName = e.target.nodeName
      if (!this.down) return
      if (nodeName !== 'rect' && nodeName !== 'svg') return

      const moveX = e.offsetX - this.startInfo.x
      const moveY = e.offsetY - this.startInfo.y
      this.props.onChange((this.props.width + moveX) / this.props.width, (this.props.height + moveY) / this.props.height)
    })
  }

  render() {
    return (<circle
      ref={x => (this.circle = x)}
      cx={this.props.cx}
      cy={this.props.cy}
      r='4'
      fill='rgba(0, 0, 0, 0)'
      style={{ cursor: 'se-resize' }}
    />)
  }
}

Circle.propTypes = {
  cx: PropTypes.number.isRequired,
  cy: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
  end: PropTypes.func.isRequired,
}

export default Circle
