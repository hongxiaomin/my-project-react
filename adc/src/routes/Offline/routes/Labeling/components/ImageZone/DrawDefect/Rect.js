/* eslint no-unused-expressions: 0, import/no-named-as-default: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import { loadImage, slicing } from 'Utils'
import { connect } from 'react-redux'
import Circle from './Circle'
import { UPDATE_SELECTED_IMAGE_INFO } from '../../../modules'

const isRight = num => (isNaN(num) || num < 0)

export class Rect extends React.Component {
  constructor() {
    super()

    this.rect = null
    this.down = false
    this.binded = false
    this.startInfo = {
      x: 0,
      y: 0,
    }
    this.state = {
      scale: {
        x: 1,
        y: 1,
      },
      offset: {
        x: 0,
        y: 0,
      },
    }
  }

  componentDidMount() {
    // 框的移动
    this.rect && this.bind()
  }

  componentWillReceiveProps() {
    if (!this.binded && this.rect) {
      this.bind()
    }

    window.addEventListener('mousemove', (e) => {
      const offset = {
        x: e.offsetX - this.startInfo.x,
        y: e.offsetY - this.startInfo.y,
      }
      const nodeName = e.target.nodeName
      if (!this.down) return
      if (nodeName !== 'rect' && nodeName !== 'svg') return

      // 禁止拖动出四边, IE11不兼容， 之后调整
      // if (
      //   (this.props.info.x + offset.x) <= 0 || // 红框是否处于大框中
      //   (this.props.info.y + offset.y) <= 0 ||
      //   (this.props.info.x + offset.x + this.props.info.width) >
      //   (this.props.imageInfo.width * this.props.aspectRatio) ||
      //   (this.props.info.y + offset.y + this.props.info.height) >
      //   (this.props.imageInfo.height * this.props.aspectRatio)
      // ) return

      this.setState({
        offset,
      })

      e.stopPropagation()
      e.preventDefault()
    })
    window.addEventListener('mouseup', () => {
      if (!this.down) return
      this.down = false

      // 更新redux
      //  defectInfo
      const { info, aspectRatio } = this.props
      const tmp = {
        x: (parseInt(info.x, 10) + parseInt(this.state.offset.x, 10)) / aspectRatio,
        y: (parseInt(info.y, 10) + parseInt(this.state.offset.y, 10)) / aspectRatio,
        endX: (parseInt(info.endX, 10) + parseInt(this.state.offset.x, 10)) / aspectRatio,
        endY: (parseInt(info.endY, 10) + parseInt(this.state.offset.y, 10)) / aspectRatio,
      }

      this.updateInfo(tmp, () => {
        this.setState({
          offset: {
            x: 0,
            y: 0,
          },
        })
      })
    })
  }

  change = (x, y) => {
    this.setState({
      scale: {
        x: x < 0 ? 0.1 : x,
        y: y < 0 ? 0.1 : y,
      },
    })
  }

  updateInfo = (info, cb) => {
    const tmp = info
    loadImage(this.props.selectedImageInfo.path)
    .then((img) => {
      const slicingURL = slicing(this.props.selectedImageInfo.path, img.width, img.height)
      tmp.preview = slicingURL(tmp.x, tmp.y, tmp.endX - tmp.x, tmp.endY - tmp.y)
    })
    .then(() => {
      const { selectedImageInfo,
        dispatch, info: propsInfo, selectedImageInfo: { defectInfo } } = this.props
      // 这个地方需要更新
      dispatch({
        type: UPDATE_SELECTED_IMAGE_INFO,
        data: {
          ...selectedImageInfo,
          defectCurrentId: propsInfo.id,
          defectInfo: defectInfo.map((item) => {
            if (item.id === propsInfo.id) {
              return {
                ...item,
                ...tmp,
              }
            }
            return item
          }),
        },
      })

      cb && cb()
    })
  }

  end = () => {
    const { info, aspectRatio } = this.props
    const tmp = {
      x: parseInt(info.x, 10) / aspectRatio,
      y: parseInt(info.y, 10) / aspectRatio,
      endX: (info.x + (info.width * this.state.scale.x)) / aspectRatio,
      endY: (info.y + (info.height * this.state.scale.y)) / aspectRatio,
    }

    this.updateInfo(tmp, () => {
      this.setState({
        scale: {
          x: 1,
          y: 1,
        },
      })
    })
  }

  bind = () => {
    this.rect.addEventListener('mousedown', (e) => {
      if (e.ctrlKey) return
      this.down = true
      this.startInfo = {
        x: e.offsetX,
        y: e.offsetY,
      }

      e.stopPropagation()
      e.preventDefault()
    })

    this.binded = true
  }

  render() {
    const { info, borderColor } = this.props
    const style = {
      cursor: 'move',
      vectorEffect: 'non-scaling-stroke',
      transform: `translate(${this.state.offset.x}px, ${this.state.offset.y}px) scale(${this.state.scale.x}, ${this.state.scale.y})`,
      transformOrigin: `${this.props.info.x}px ${this.props.info.y}px`,
    }

    // style.transformOrigin = `${info.x}px ${info.y}px`
    if (isRight(info.x) || isRight(info.y) || isRight(info.width) || isRight(info.height)) {
      return null
    }

    // 文字的坐标
    let textX
    let textY
    const textStyle = {
      fontSize: 18,
      transform: `translate(${this.state.offset.x}px, ${this.state.offset.y}px)`,
      transformOrigin: `${this.props.info.x}px ${this.props.info.y}px`,
    }
    if (info.width > info.height) {
      textX = info.x
      textY = info.y < 25 ? (info.y + info.height + 20) : info.y - 7
    } else {
      textX = info.x < 25 ? (info.x + info.width + 17) : info.x - 15
      textY = info.y
      textStyle.writingMode = 'tb'
    }

    return (<g>
      {info.defectCode
        ? <text
          x={textX}
          y={textY}
          fill={borderColor}
          style={textStyle}
        >{info.defectCode}</text>
        : ''
      }
      <rect
        ref={x => (this.rect = x)}
        stroke={borderColor}
        x={info.x}
        y={info.y}
        width={info.width}
        height={info.height}
        strokeWidth={2 / this.props.scale}
        fill="rgba(0, 0, 0, 0)"
        style={style}
      />
      <Circle
        cx={this.props.info.endX}
        cy={this.props.info.endY}
        width={info.width}
        height={info.height}
        onChange={this.change}
        end={this.end}
      />
    </g>)
  }
}

Rect.propTypes = {
  borderColor: PropTypes.string.isRequired,
  info: PropTypes.object.isRequired,
  scale: PropTypes.number.isRequired,
  aspectRatio: PropTypes.number.isRequired,
  dispatch: PropTypes.func.isRequired,
  imageInfo: PropTypes.object.isRequired,
  selectedImageInfo: PropTypes.object.isRequired,
}

export default connect(state => ({
  scale: state.labelingTool.controlPanel.zoom,
  selectedImageInfo: state.labelingTool.selectedImageInfo,
}))(Rect)
