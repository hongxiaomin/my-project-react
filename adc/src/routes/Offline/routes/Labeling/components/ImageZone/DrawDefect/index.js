import DrawDefect from './DrawDefect'
export default DrawDefect
