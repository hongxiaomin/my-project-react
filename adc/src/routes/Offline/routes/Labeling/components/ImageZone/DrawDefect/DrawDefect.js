
/* eslint import/no-named-as-default: 0 */
import React from 'react'
import { isEmpty } from 'lodash'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { loadImage, slicing } from 'Utils'
import Rect from './Rect'
import { UPDATE_SELECTED_IMAGE_INFO } from '../../../modules'

export class DrawDefect extends React.Component {

  constructor(props) {
    super(props)

    this.drawDefect = null
    this.svg = null
    this.down = false
    this.moved = false
    this.binded = false
    this.startInfo = {
      x: 0,
      y: 0,
    }

    this.state = {
      rect: {}, // 新增的时候， 保存信息
    }
  }

  componentDidMount() {
    this.bind()
  }

  componentWillReceiveProps() {
    if (!this.binded) {
      this.bind()
    }
  }

  getAddLabeling = (event, aspectRatio = 1) => {
    const info = {
      id: Symbol(Date.now()),
    }
    const { x: startX, y: startY } = this.startInfo

    // 两种情况
    if (startX < event.offsetX) {
      info.x = startX / aspectRatio
      info.endX = event.offsetX / aspectRatio
    } else {
      info.x = event.offsetX / aspectRatio
      info.endX = startX / aspectRatio
    }

    if (startY < event.offsetY) {
      info.y = startY / aspectRatio
      info.endY = event.offsetY / aspectRatio
    } else {
      info.y = event.offsetY / aspectRatio
      info.endY = startY / aspectRatio
    }

    const { selectedDefaultDefect } = this.props
    if (!isEmpty(selectedDefaultDefect)) {
      info.defectId = selectedDefaultDefect.defectId
      info.defectCode = selectedDefaultDefect.defectCode
    }
    return info
  }

  bind = () => {
    this.drawDefect.addEventListener('mousedown', (e) => {
      if (e.ctrlKey) return
      /**
       * 记录初始化位置
       * 1. 从浏览器左上角开始计算， 要加入下拉框位置
       */

      this.startInfo = { x: e.offsetX, y: e.offsetY }
      this.down = true
      this.moved = false
    })

    this.drawDefect.addEventListener('mousemove', this.move)

    window.addEventListener('mouseup', (e) => {
      if (!this.down) return
      if (!this.moved) {
        this.down = false
        return
      }
      // 考虑在框外部释放鼠标的可能性
      this.down = false

      const { selectedImageInfo, imageInfo, dispatch } = this.props
      loadImage(selectedImageInfo.path)
        .then((img) => {
          const slicingURL = slicing(selectedImageInfo.path, img.width, img.height)
          const tmp = this.getAddLabeling(e, this.props.imageInfo.aspectRatio)
          tmp.preview = slicingURL(tmp.x, tmp.y, tmp.endX - tmp.x, tmp.endY - tmp.y)
          return tmp
        })
        .then((tmp) => {
          // 新增一条数据， 查找所有的数据是否已经都有defectCode, 如果没有， 删掉
          const defectInfo = []
          selectedImageInfo.defectInfo.forEach((item) => {
            if (!item.defectCode) return
            defectInfo.push(item)
          })
          dispatch({
            type: UPDATE_SELECTED_IMAGE_INFO,
            data: {
              ...selectedImageInfo,
              defectCurrentId: tmp.id,
              defectInfo: defectInfo.map(item => ({
                ...item,
                x: item.x / imageInfo.aspectRatio,
                y: item.y / imageInfo.aspectRatio,
                endX: item.endX / imageInfo.aspectRatio,
                endY: item.endY / imageInfo.aspectRatio,
              })).concat(tmp),
            },
          })
          this.setState({
            rect: {},
          })
        })
    })

    this.binded = true
  }

  move = (e) => {
    if (!this.down) return
    if (this.startInfo.x !== e.offsetX || this.startInfo.y !== e.offsetY) {
      this.moved = true
    }
    this.setState({
      rect: this.getAddLabeling(e),
    })
  }

  render() {
    let Rects = []
    const styles = {
      position: 'absolute',
      left: 0,
      top: 0,
      width: '100%',
      height: '100%',
    }
    const {
      selectedImageInfo: {
        defectInfo = [],
        defectCurrentId,
      },
    } = this.props

    Rects = [...defectInfo]
    if (!isEmpty(this.state.rect)) {
      Rects.push(this.state.rect)
    }

    return (<div
      ref={(x) => { this.drawDefect = x }}
      style={styles}
    >
      <svg
        ref={(c) => { this.svg = c }}
        width='100%'
        height='100%'
        preserveAspectRatio="xMinYMin meet"
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <title>Labeling Tool</title>
        {Rects.map(item => (<Rect
          key={item.id.toString()}
          aspectRatio={this.props.imageInfo.aspectRatio}
          imageInfo={this.props.imageInfo}
          info={{
            id: item.id,
            x: item.x,
            y: item.y,
            endX: item.endX,
            endY: item.endY,
            width: item.endX - item.x,
            height: item.endY - item.y,
            defectCode: item.defectCode,
          }}
          borderColor={(defectCurrentId === item.id) ? 'red' : '#FF9B98'}
        />))}
      </svg>
    </div>)
  }
}

DrawDefect.propTypes = {
  imageInfo: PropTypes.object.isRequired,
  selectedImageInfo: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
  selectedDefaultDefect: PropTypes.object.isRequired,
}

export default connect((state, ownProps) => {
  const { labelingTool } = state
  const selectedImageInfo = labelingTool.selectedImageInfo
  const selectedDefaultDefect = labelingTool.selectedDefaultDefect
  let info = []

  if (selectedImageInfo && selectedImageInfo.defectInfo && selectedImageInfo.defectInfo.length) {
    info = selectedImageInfo.defectInfo.map(item => ({
      ...item,
      x: item.x * ownProps.imageInfo.aspectRatio,
      y: item.y * ownProps.imageInfo.aspectRatio,
      endX: item.endX * ownProps.imageInfo.aspectRatio,
      endY: item.endY * ownProps.imageInfo.aspectRatio,
    }))
  }

  return {
    selectedDefaultDefect,
    selectedImageInfo: {
      ...labelingTool.selectedImageInfo,
      defectInfo: info,
    },
  }
})(DrawDefect)
