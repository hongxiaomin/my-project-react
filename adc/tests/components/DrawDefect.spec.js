import React from 'react'
import { shallow, mount } from 'enzyme'
import { DrawDefect } from 'routes/Offline/routes/Labeling/components/ImageZone/DrawDefect/DrawDefect'

describe('(Component) DrawDefect', () => {
  const scale = 0.5
  const offset = {}
  const div = {}
  const imageInfo = {}
  const selectedImageInfo = {}
  const addLabeling = {}
  const dispatch = jest.fn()
  const selectedDefectCode = {}
  it('render normally', () => {
    const wrapper = shallow(
      <DrawDefect
        scale={scale}
        offset={offset}
        div={div}
        imageInfo={imageInfo}
        selectedImageInfo={selectedImageInfo}
        addLabeling={addLabeling}
        dispatch={dispatch}
        selectedDefectCode={selectedDefectCode}
      />,
    )
    // const nativeEvent = { nativeEvent: { stopImmediatePropagation: _.noop } }
    // wrapper.find('Modal').simulate('click', nativeEvent)
    // expect(wrapper.state('loading')).toBe(false)
    expect(wrapper).toMatchSnapshot()
  })
  it('render mount', () => {
    const wrapper = mount(
      <DrawDefect
        scale={scale}
        offset={offset}
        div={div}
        imageInfo={imageInfo}
        selectedImageInfo={selectedImageInfo}
        addLabeling={addLabeling}
        dispatch={dispatch}
        selectedDefectCode={selectedDefectCode}
      />,
    )
    // const nativeEvent = { nativeEvent: { stopImmediatePropagation: _.noop } }
    // wrapper.find('Modal').simulate('click', nativeEvent)
    // expect(wrapper.state('loading')).toBe(false)
    expect(wrapper).toMatchSnapshot()
  })
})

