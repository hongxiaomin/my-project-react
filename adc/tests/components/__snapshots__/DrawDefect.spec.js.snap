// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) DrawDefect render mount 1`] = `
<DrawDefect
  addLabeling={Object {}}
  dispatch={[Function]}
  div={Object {}}
  imageInfo={Object {}}
  offset={Object {}}
  scale={0.5}
  selectedDefectCode={Object {}}
  selectedImageInfo={Object {}}
>
  <div
    style={
      Object {
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <svg
      height="100%"
      preserveAspectRatio="xMinYMin meet"
      width="100%"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <title>
        Labeling Tool
      </title>
    </svg>
  </div>
</DrawDefect>
`;

exports[`(Component) DrawDefect render normally 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": "100%",
    }
  }
>
  <svg
    height="100%"
    preserveAspectRatio="xMinYMin meet"
    width="100%"
    xmlns="http://www.w3.org/2000/svg"
    xmlnsXlink="http://www.w3.org/1999/xlink"
  >
    <title>
      Labeling Tool
    </title>
  </svg>
</div>
`;
